File: Bank/cpp/tests/AccountShould.cpp
#include <gmock/gmock.h>
using namespace ::testing;
#include "Account.h"
#include "MockTransactionRepository.h"
#include "Transaction.h"
#include "MockStatementPrinter.h"
#include <memory>
#include <list>

class AccountShould : public Test
{
private:
        using MockTransactionRepository = NiceMock<MockTransactionRepository>;
        using MockStatementPrinter = NiceMock<MockStatementPrinter>;
protected:
        std::shared_ptr<MockTransactionRepository> transactionRepository{new MockTransactionRepository};
        std::shared_ptr<MockStatementPrinter> statementPrinter{new MockStatementPrinter};
        Account account{transactionRepository, statementPrinter};
};

TEST_F(AccountShould, StoreDepositTransaction)
{
        EXPECT_CALL(*transactionRepository, addDeposit(100)).Times(1);
        account.deposit(100);
}

TEST_F(AccountShould, StoreWithdrawalTransaction)
{
        EXPECT_CALL(*transactionRepository, addWithdrawal(100)).Times(1);
        account.withdraw(100);
}

TEST_F(AccountShould, PrintStatementOfAccount)
{
        std::list<Transaction> transactions;
        EXPECT_CALL(*transactionRepository, allTransactions()).WillOnce(Return(transactions));
        EXPECT_CALL(*statementPrinter, print(transactions)).Times(1);
        account.printStatement();
}

File: Bank/java/src/test/java/BankKata/AccountShould.java
package BankKata;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collections;
import java.util.List;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class AccountShould {

    @Mock private TransactionRepository transactionRepository;
    @Mock private StatementPrinter statementPrinter;
    private Account account;

    @Before
    public void setUp() {
        this.account = new Account(transactionRepository, statementPrinter);
    }

    @Test
    public void store_a_deposit_transaction() {
        TransactionAmount amount = new TransactionAmount(100);
        this.account.deposit(amount);
        verify(transactionRepository).addDeposit(amount);
    }

    @Test
    public void store_a_withdrawal_transaction() {
        TransactionAmount amount = new TransactionAmount(100);
        this.account.withdraw(amount);
        verify(transactionRepository).addWithdrawal(amount);
    }

    @Test
    public void print_statement_of_transactions() {
        List<Transaction> transactions = Collections.emptyList();
        given(transactionRepository.allTransactions()).willReturn(transactions);
        this.account.printStatement();
        verify(statementPrinter).print(transactions);
    }
}